File: scripts/diagnostics/test-console-ws.js
#!/usr/bin/env node
/**
 * Test WebSocket console endpoint
 * 1. Login to get access token
 * 2. Get VM UUID
 * 3. Call /api/vms/{uuid}/console to get ws_url
 * 4. Test WebSocket connection
 */

const http = require('http');
const https = require('https');
const { WebSocket } = require('ws');

const BASE_URL = 'http://localhost:18443';
const PASSWORDS = ['0625', 'admin', 'password'];

function httpRequest(url, options = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const client = urlObj.protocol === 'https:' ? https : http;
    
    const req = client.request(url, {
      method: options.method || 'GET',
      headers: options.headers || {},
      ...options
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          resolve({ status: res.statusCode, data: json });
        } catch (e) {
          resolve({ status: res.statusCode, data: data });
        }
      });
    });
    
    req.on('error', reject);
    
    if (options.body) {
      req.write(JSON.stringify(options.body));
    }
    
    req.end();
  });
}

async function login(username = 'admin', password = null) {
  const passwords = password ? [password] : PASSWORDS;
  
  for (const pwd of passwords) {
    try {
      const result = await httpRequest(`${BASE_URL}/api/auth/login`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: { username, password: pwd }
      });
      
      if (result.status === 200 && result.data.access_token) {
        return result.data.access_token;
      }
    } catch (e) {
      // Continue to next password
    }
  }
  
  return null;
}

async function getVmUuid(token) {
  try {
    const result = await httpRequest(`${BASE_URL}/api/vms`, {
      headers: { 'Authorization': `Bearer ${token}` }
    });
    
    if (result.status === 200 && Array.isArray(result.data) && result.data.length > 0) {
      return result.data[0].uuid;
    }
  } catch (e) {
    console.error('Error getting VM list:', e.message);
  }
  
  return null;
}

async function getConsoleUrl(token, vmUuid) {
  try {
    const result = await httpRequest(`${BASE_URL}/api/vms/${vmUuid}/console`, {
      headers: { 'Authorization': `Bearer ${token}` }
    });
    
    if (result.status === 200) {
      return {
        ws_url: result.data.ws_url,
        protocol: result.data.protocol,
        expires_at: result.data.expires_at
      };
    } else {
      console.error(`ERROR: Failed to get console URL: ${result.status}`);
      console.error(`Response: ${JSON.stringify(result.data)}`);
      return null;
    }
  } catch (e) {
    console.error('Error getting console URL:', e.message);
    return null;
  }
}

function testWebSocket(wsUrl) {
  return new Promise((resolve) => {
    console.log(`\n[WS TEST] Connecting to: ${wsUrl}`);
    
    const ws = new WebSocket(wsUrl, {
      handshakeTimeout: 5000,
      perMessageDeflate: false
    });
    
    let opened = false;
    let receivedMessage = false;
    let closed = false;
    
    const timeout = setTimeout(() => {
      if (!closed) {
        console.log('[WS TEST] ✓ Connection timeout (5s) - connection is alive');
        ws.close();
      }
    }, 5000);
    
    ws.on('open', () => {
      opened = true;
      console.log('[WS TEST] ✓ Connection opened successfully');
    });
    
    ws.on('message', (data) => {
      receivedMessage = true;
      const preview = data.toString().slice(0, 80);
      const type = Buffer.isBuffer(data) ? 'Buffer' : typeof data;
      console.log(`[WS TEST] ✓ Received message (type: ${type}, len: ${data.length}): ${preview}`);
    });
    
    ws.on('error', (error) => {
      console.log(`[WS TEST] ✗ Error: ${error.message}`);
      clearTimeout(timeout);
      resolve(false);
    });
    
    ws.on('close', (code, reason) => {
      closed = true;
      clearTimeout(timeout);
      
      if (opened) {
        if (receivedMessage) {
          console.log(`[WS TEST] ✓ Connection closed normally: code=${code}, reason=${reason || 'none'}`);
          resolve(true);
        } else {
          console.log(`[WS TEST] ⚠ Connection closed without receiving data: code=${code}, reason=${reason || 'none'}`);
          // Still consider it a success if connection was opened
          resolve(code === 1000 || code === 1001);
        }
      } else {
        console.log(`[WS TEST] ✗ Connection closed immediately: code=${code}, reason=${reason || 'none'}`);
        resolve(false);
      }
    });
  });
}

async function main() {
  console.log('=== Console WebSocket Test ===\n');
  
  // Step 1: Login
  console.log('[1] Logging in...');
  const token = await login();
  if (!token) {
    console.log('✗ Login failed');
    process.exit(1);
  }
  console.log(`✓ Login successful (token: ${token.substring(0, 20)}...)`);
  
  // Step 2: Get VM UUID
  console.log('\n[2] Getting VM list...');
  const vmUuid = await getVmUuid(token);
  if (!vmUuid) {
    console.log('✗ No VM found');
    process.exit(1);
  }
  console.log(`✓ Found VM: ${vmUuid}`);
  
  // Step 3: Get console URL
  console.log('\n[3] Getting console URL...');
  const consoleInfo = await getConsoleUrl(token, vmUuid);
  if (!consoleInfo || !consoleInfo.ws_url) {
    console.log('✗ Failed to get console URL');
    process.exit(1);
  }
  console.log(`✓ Console URL received:`);
  console.log(`  ws_url: ${consoleInfo.ws_url}`);
  console.log(`  protocol: ${consoleInfo.protocol}`);
  console.log(`  expires_at: ${consoleInfo.expires_at}`);
  
  // Step 4: Test WebSocket
  console.log('\n[4] Testing WebSocket connection...');
  const success = await testWebSocket(consoleInfo.ws_url);
  
  if (success) {
    console.log('\n✓✓✓ ALL TESTS PASSED ✓✓✓');
    process.exit(0);
  } else {
    console.log('\n✗✗✗ WEBSOCKET TEST FAILED ✗✗✗');
    process.exit(1);
  }
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: scripts/diagnostics/test-console-ws.py
#!/usr/bin/env python3
"""
Test WebSocket console endpoint
1. Login to get access token
2. Get VM UUID
3. Call /api/vms/{uuid}/console to get ws_url
4. Test WebSocket connection
"""
import sys
import json
import requests
import asyncio
import websockets
from urllib.parse import urlparse, parse_qs

BASE_URL = "http://localhost:18443"
# Try different passwords
PASSWORDS = ["0625", "admin", "password"]

def login(username="admin", password=None):
    """Login and get access token"""
    if password is None:
        for pwd in PASSWORDS:
            resp = requests.post(
                f"{BASE_URL}/api/auth/login",
                json={"username": username, "password": pwd},
                timeout=5
            )
            if resp.status_code == 200:
                data = resp.json()
                return data.get("access_token")
        return None
    
    resp = requests.post(
        f"{BASE_URL}/api/auth/login",
        json={"username": username, "password": password},
        timeout=5
    )
    if resp.status_code == 200:
        data = resp.json()
        return data.get("access_token")
    return None

def get_vm_uuid(token):
    """Get first VM UUID"""
    headers = {"Authorization": f"Bearer {token}"}
    resp = requests.get(f"{BASE_URL}/api/vms", headers=headers, timeout=5)
    if resp.status_code == 200:
        vms = resp.json()
        if vms and len(vms) > 0:
            return vms[0].get("uuid")
    return None

def get_console_url(token, vm_uuid):
    """Get console WebSocket URL"""
    headers = {"Authorization": f"Bearer {token}"}
    resp = requests.get(
        f"{BASE_URL}/api/vms/{vm_uuid}/console",
        headers=headers,
        timeout=5
    )
    if resp.status_code == 200:
        data = resp.json()
        return data.get("ws_url"), data.get("protocol"), data.get("expires_at")
    else:
        print(f"ERROR: Failed to get console URL: {resp.status_code}")
        print(f"Response: {resp.text}")
        return None, None, None

async def test_websocket(ws_url):
    """Test WebSocket connection"""
    print(f"\n[WS TEST] Connecting to: {ws_url}")
    
    try:
        # Parse URL to extract origin if needed
        parsed = urlparse(ws_url)
        origin = f"{parsed.scheme.replace('ws', 'http')}://{parsed.netloc}"
        
        async with websockets.connect(
            ws_url,
            origin=origin if parsed.scheme == "wss" else None,
            ping_interval=None,
            ping_timeout=None,
            close_timeout=5
        ) as ws:
            print("[WS TEST] ✓ Connection opened successfully")
            
            # Wait for messages or close
            try:
                # Wait up to 3 seconds for any message
                message = await asyncio.wait_for(ws.recv(), timeout=3.0)
                msg_preview = str(message)[:80] if len(str(message)) > 80 else str(message)
                print(f"[WS TEST] ✓ Received message (type: {type(message).__name__}, len: {len(str(message))}): {msg_preview}")
            except asyncio.TimeoutError:
                print("[WS TEST] ✓ No message received (timeout after 3s) - connection is alive")
            except websockets.exceptions.ConnectionClosed as e:
                print(f"[WS TEST] ✗ Connection closed by server: code={e.code}, reason={e.reason}")
                return False
            
            # Send a test message (if VNC protocol supports it)
            # For VNC, we might not send anything, just keep connection alive
            await asyncio.sleep(1)
            
            print("[WS TEST] ✓ WebSocket test PASSED - connection is stable")
            return True
            
    except websockets.exceptions.InvalidStatusCode as e:
        print(f"[WS TEST] ✗ Connection failed: HTTP {e.status_code}")
        if hasattr(e, 'headers'):
            print(f"  Headers: {e.headers}")
        return False
    except websockets.exceptions.ConnectionClosedError as e:
        print(f"[WS TEST] ✗ Connection closed immediately: code={e.code}, reason={e.reason}")
        return False
    except Exception as e:
        print(f"[WS TEST] ✗ Error: {type(e).__name__}: {e}")
        return False

def main():
    print("=== Console WebSocket Test ===")
    
    # Step 1: Login
    print("\n[1] Logging in...")
    token = login()
    if not token:
        print("✗ Login failed")
        sys.exit(1)
    print(f"✓ Login successful (token: {token[:20]}...)")
    
    # Step 2: Get VM UUID
    print("\n[2] Getting VM list...")
    vm_uuid = get_vm_uuid(token)
    if not vm_uuid:
        print("✗ No VM found")
        sys.exit(1)
    print(f"✓ Found VM: {vm_uuid}")
    
    # Step 3: Get console URL
    print("\n[3] Getting console URL...")
    ws_url, protocol, expires_at = get_console_url(token, vm_uuid)
    if not ws_url:
        print("✗ Failed to get console URL")
        sys.exit(1)
    print(f"✓ Console URL received:")
    print(f"  ws_url: {ws_url}")
    print(f"  protocol: {protocol}")
    print(f"  expires_at: {expires_at}")
    
    # Step 4: Test WebSocket
    print("\n[4] Testing WebSocket connection...")
    success = asyncio.run(test_websocket(ws_url))
    
    if success:
        print("\n✓✓✓ ALL TESTS PASSED ✓✓✓")
        sys.exit(0)
    else:
        print("\n✗✗✗ WEBSOCKET TEST FAILED ✗✗✗")
        sys.exit(1)

if __name__ == "__main__":
    main()
